moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp

```diff
@@ -0,0 +1,141 @@
+ *     with the distribution.
+ *   * Neither the name of PickNik Inc. nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Sebastian Jahr */
+
+#include "get_group_urdf_capability.h"
+
+#include <moveit/moveit_cpp/moveit_cpp.h>
+#include <moveit/move_group/capability_names.h>
+#include <moveit/utils/logger.hpp>
+#include <tinyxml2.h>
+
+namespace move_group
+{
+
+namespace
+{
+const auto LOG_NAME = std::string("GetUrdfService");
+rclcpp::Logger getLogger()
+{
+  return moveit::getLogger(LOG_NAME);
+}
+const auto JOINT_ELEMENT_CLOSING = std::string("</joint>");
+const auto LINK_ELEMENT_CLOSING = std::string("</link>");
+const auto ROBOT_ELEMENT_CLOSING = std::string("</robot>");
+}  // namespace
+
+GetUrdfService::GetUrdfService() : MoveGroupCapability("get_group_urdf")
+{
+}
+
+void GetUrdfService::initialize()
+{
+  get_urdf_service_ = context_->moveit_cpp_->getNode()->create_service<moveit_msgs::srv::GetGroupUrdf>(
+      GET_URDF_SERVICE_NAME,
+      [this](const std::shared_ptr<moveit_msgs::srv::GetGroupUrdf::Request>& req,
+             const std::shared_ptr<moveit_msgs::srv::GetGroupUrdf::Response>& res) {
+        res->error_code.source = LOG_NAME;
+        const auto subgroup = context_->moveit_cpp_->getRobotModel()->getJointModelGroup(req->group_name);
+        // Check if group exists in loaded robot model
+        if (!subgroup)
+        {
+          const auto error_string = std::string("Cannot create URDF because planning group '") + req->group_name +
+                                    std::string("' does not exist");
+          RCLCPP_ERROR(getLogger(), "%s", error_string.c_str());
+          res->error_code.message = error_string;
+          res->error_code.val = moveit_msgs::msg::MoveItErrorCodes::FAILURE;
+          return;
+        }
+        // Get robot description string
+        std::string full_urdf_string;
+        context_->moveit_cpp_->getNode()->get_parameter_or("robot_description", full_urdf_string, std::string(""));
```

Comment:
Doesn't urdfdom help us with safely parsing the URDF from the RobotModel? It seems weird that we have utility libraries for managing URDF, but we have to build our custom parser.

Reply:
Good point, I can use [exportURDF](https://github.com/ros/urdfdom/blob/29426ec119e9c3a9df9fc705d08c262a97b7f285/urdf_parser/src/model.cpp#L312) and then use [accept](https://leethomason.github.io/tinyxml2/classtinyxml2_1_1_x_m_l_document.html#a853f05e2f7f51eb598e32ec61d8bc53c) to print the text. I will try this tomorrow morning

Reply:
Nope that did not work :disappointed: Keep using the parameter API for that